File: library/src/main/java/org/fbme/lib/common/CompositeReference.java
package org.fbme.lib.common;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public interface CompositeReference<DeclarationPathT extends DeclarationPath> {

    @Nullable DeclarationPathT getTarget();

    @NotNull List<Identifier> getPath();

    @NotNull String getPresentation();

    void setPath(@NotNull List<Identifier> path);

    default void setTarget(@NotNull DeclarationPathT target) {
        setPath(target.getDeclarations().stream().map(Declaration::getIdentifier).collect(Collectors.toList()));
    }

    default void setTargetNames(@NotNull List<String> names) {
        setPath(names.stream().map(StringIdentifier::new).collect(Collectors.toList()));
    }

    default void setFQName(@NotNull String fqName) {
        setTargetNames(Arrays.asList(fqName.split("\\.")));
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/common/CompositeReference.kt
package org.fbme.lib.common

interface CompositeReference<DeclarationPathT : DeclarationPath> {
    fun getTarget(): DeclarationPathT?
    fun getPath(): List<Identifier>
    val presentation: String
    fun setPath(path: List<Identifier>)
    fun setTarget(target: DeclarationPathT) {
        setPath(target.declarations.map { obj: Declaration -> obj.identifier })
    }

    fun setTargetNames(names: List<String>) {
        setPath(names.map { value -> StringIdentifier(value) })
    }

    fun setFQName(fqName: String) {
        setTargetNames(listOf(*fqName.split(".").toTypedArray()))
    }
}
